src/utils/include/utils/tensor.hpp

```diff
@@ -0,0 +1,128 @@
+#ifndef TENSOR_HPP
+#define TENSOR_HPP
+
+#include <algorithm>
+#include <array>
+#include <numeric>
+#include <vector>
+
+#include <boost/range/algorithm/transform.hpp>
+#include <boost/range/numeric.hpp>
+
+namespace Utils {
+
+/**
+ * @brief Heap-allocated tensor class.
+ * @tparam T data type
+ */
+template <typename T> class Tensor {
+public:
+  using value_type = T;
+  using reference = T &;
+  using const_reference = T const &;
+  using pointer = T *;
+  using const_pointer = const value_type *;
+  using iterator = typename std::vector<T>::iterator;
+  using const_iterator = typename std::vector<T>::const_iterator;
+
+  /**
+   * @brief Creates a tensor with given extents
+   * @param extents Size for each dimension
+   */
+  explicit Tensor(std::initializer_list<std::size_t> const &extents)
+      : Tensor(std::begin(extents), std::end(extents)) {}
+
+  template <typename Iter>
+  Tensor(Iter extents_begin, Iter extents_end)
+      : m_extents(extents_begin, extents_end) {
+    m_data = std::vector<T>(std::accumulate(m_extents.begin(), m_extents.end(),
+                                            1, std::multiplies<std::size_t>()));
+
+    boost::transform(m_extents, std::back_inserter(m_strides),
+                     [stride = std::size_t(1)](std::size_t dim) mutable {
+                       auto const old = stride;
+                       stride *= dim;
+
+                       return old;
+                     });
+  }
+
+  /**
+   * @brief Element access.
+   * @param indices index for each dimension
+   */
+  template <typename... size_ts>
+  const_reference operator()(size_ts... indices) const {
+    return operator()({static_cast<std::size_t>(indices)...});
+  }
+
+  template <typename... size_ts> reference operator()(size_ts... indices) {
+    return operator()({static_cast<std::size_t>(indices)...});
+  }
+
+  const_reference
+  operator()(std::initializer_list<std::size_t> const &indices) const {
+    validate_indices(indices);
```

Comment:
I'm not sure that the default access should be range checked, this is not e.g. like `std::array` or `std::vector` behave. Here there is a special penalty to it because you need to load additional unrelated heap memory that you would not need otherwise. I think this should be an `assert`, and you should provide an `at` member that does a range checked version so that the caller can decide.